nautobot_circuit_maintenance/views.py

```diff
@@ -299,3 +314,76 @@ def get(self, request, *args, **kwargs):
                 "source_type": source.__class__.__name__,
             },
         )
+
+
+def google_authorize(request, slug):
+    """View to start the Google OAuth authorization flow."""
+    # Create flow instance to manage the OAuth 2.0 Authorization Grant Flow steps.
+    notification_source = models.NotificationSource.objects.get(slug=slug)
+    source = Source.init(name=notification_source.name)
+    request.session["CLIENT_SECRETS_FILE"] = source.credentials_file
+    request.session["SCOPES"] = source.SCOPES
+    request.session["SOURCE_SLUG"] = slug
+
+    # Create flow instance to manage the OAuth 2.0 Authorization Grant Flow steps.
+    flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(
+        request.session["CLIENT_SECRETS_FILE"], scopes=request.session["SCOPES"]
+    )
+    # The URI created here must exactly match one of the authorized redirect URIs
+    # for the OAuth 2.0 client, which you configured in the API Console. If this
+    # value doesn't match an authorized URI, you will get a 'redirect_uri_mismatch'
+    # error.
+
+    flow.redirect_uri = (
+        request.scheme
+        + "://"
+        + request.get_host()
+        + reverse("plugins:nautobot_circuit_maintenance:google_oauth2callback")
+    )
+    authorization_url, state = flow.authorization_url(
+        # Enable offline access so that you can refresh an access token without
+        # re-prompting the user for permission. Recommended for web server apps.
+        access_type="offline",
+        # Enable incremental authorization. Recommended as a best practice.
+        include_granted_scopes="true",
+        # Asking for consent everytime is providing a refresh token that if not used only happens
+        # the first time.
+        prompt="consent",
+    )
+    # Store the state so the callback can verify the auth server response.
+    request.session["state"] = state
+
+    return redirect(authorization_url)
+
+
+def google_oauth2callback(request):
+    """View to receive the callback from Google OAuth authorization flow."""
+    # Specify the state when creating the flow in the callback so that it can
+    # verified in the authorization server response.
+    state = request.session["state"]
+    flow = google_auth_oauthlib.flow.Flow.from_client_secrets_file(
+        request.session["CLIENT_SECRETS_FILE"], scopes=request.session["SCOPES"], state=state
```

Comment:
From a security standpoint, do we need to check that CLIENT_SECRETS_FILE is associated with a defined Source, or anything like that? I'm trying to understand whether this could potentially allow access to files that shouldn't be accessible or anything like that.

Reply:
do you mean by an internally manipulated session? 
at the end, if the defined info in the file is not matching the info that Google is authorizing via the callback, it should fail.
Will think a bit more to see any potential problem

Reply:
Since this is a View, any client (browser, Postman, etc.) could send a request to this endpoint - not just the Google server. Just want to make sure we don't have some sort of risk of spoofing here.

Reply:
I understand your concern, but the security problem is covered by the "state" info injected in the `google_authorize' view that it's verified in the authorization server response.
Obviously, as this is an open endpoint, what could happen is that if the session doesn't contain the expected keys: "state", "CLIENT_SECRETS_FILE", "SCOPES" and "SOURCE_SLUG", it will raise a `KeyError` exception, so I would handle this gracefully with a simple `get`